### ISSUE TITLE
"unreachable_code" warns even on "unreachable!" itself

### ISSUE THREAD

**Issue by safinaskar** [rust-lang/rust#107953]
### Code

```Rust
pub async fn get(client: &reqwest::Client, params: &[(&str, &str)]) -> String {
    loop {
        let response = client.get("https://example.com")
            .query(params).send().await.unwrap();
        if response.status().is_success() {
            return response.text().await.unwrap();
        }
        assert!(response.status() == reqwest::StatusCode::TOO_MANY_REQUESTS);
    }
    unreachable!();
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
warning: unreachable statement
  --> src/lib.rs:10:5
   |
2  | /     loop {
3  | |         let response = client.get("https://example.com")
4  | |             .query(params).send().await.unwrap();
5  | |         if response.status().is_success() {
...  |
8  | |         assert!(response.status() == reqwest::StatusCode::TOO_MANY_REQUESTS);
9  | |     }
   | |_____- any code following this expression is unreachable
10 |       unreachable!();
   |       ^^^^^^^^^^^^^^ unreachable statement
   |
   = note: `#[warn(unreachable_code)]` on by default
   = note: this warning originates in the macro `unreachable` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 1.11s
```

### Desired output

```Shell
There should be no warning
```

### Rationale and extra context

Code above is based on my actual production code. The intention is this: continue issuing requests as long as we get TOO_MANY_REQUESTS. When we get SUCCESS, return from the function. I put `unreachable!` to the end intentionally: to make sure that that line is actually unreachable. I. e. I want to get panic if I'm wrong.

In this particular example rustc can prove that that piece is unreachable and thus give me a warning. But I don't like it. I don't want to remove `unreachable!`. It is present for a reason.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=684eaca4cc578206f971cc6345e3d0c6

Here is even more reduced version: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e90f7dbeb04fd15ed2ae6354cb158fce (so that you can see that async and reqwest are not related here)

**Comment by eggyal** [comment]
Surely then this is a perfect use-case for applying `#[allow(unreachable_code)]` to the `unreachable!()` call?

**Comment by safinaskar** [comment]
@eggyal , you mean I should place `#[allow(unreachable_code)]` in my code? Well, any `allow` constructs look dangerous for me. They all scream: "Hey! Dragons here! Here we are doing something dangerous and we disable some compiler checks, so that the compiler doesn't tell us about this. So **we** (not rustc) are now have responsibility to check everything is okay"

**Comment by eggyal** [comment]
I think it's more useful to have the compiler warn that an `unreachable!` invocation is in fact guaranteed to be unreachable and therefore can be removed than it is to have such warnings suppressed by default, especially since it's trivial to manually suppress the warning in rare cases such as these where it's desired.  But that's just my humble opinion, perhaps the team will see it differently.

**Comment by ChrisDenton** [comment]
Hm, in this particular case you can have a compile time guarantee that code after the loop can never be reached because the function signature requires returning a value other than `()`. You just need to add a semicolon after the loop.

But in the general case it does seem like it would be useful to have something similar to `compile_error` but that's removed for probably unreachable code.

**Comment by safinaskar** [comment]
@eggyal
> and therefore can be removed

I don't want to remove it. What if I change something and this place becomes reachable?

Also, if the team thinks that this code should give a warning, then I want some other macro (say, `static_unreachable!`), which will check statically that code is unreachable. (If someone can give implementation, this will be nice.)

@ChrisDenton
> Hm, in this particular case you can have a compile time guarantee that code after the loop can never be reached because the function signature requires returning a value other than ()

Yes. But what if in the future I change function return value?

**Comment by ChrisDenton** [comment]
> @eggyal
> > and therefore can be removed
> 
> I don't want to remove it. What if I change something and this place becomes reachable?
> 
> Also, if the team thinks that this code should give a warning, then I want some other macro (say, `static_unreachable!`), which will check statically that code is unreachable. (If someone can give implementation, this will be nice.)

I think you misunderstood me. `compile_error!(()` always causes a compile error, even if the code path can never be reached. I was suggesting a variant of that which doesn't error if the macro provably can't be reached (I.e. it's as if the `compile_error!()` were removed). So `static_unreachable` was exactly what I was suggesting.

> Yes. But what if in the future I change function return value?

I was agreeing with that, hence the paragraph that followed.

**Comment by eggyal** [comment]
If this is to be remedied, rather than special case the `unreachable_code` lint it may be better to include an `#[allow(unreachable_code)]` attribute in the `unreachable!` macro's expansion.

**Comment by timotree3** [comment]
Using the unstable `#[expect()]` feature, you can get a compile error if the code is reachable, regardless of the return type of the function 

```rust
#![feature(lint_reasons)]

pub fn reachable() {
    loop { break; }
    #[deny(unfulfilled_lint_expectations)]
    #[expect(unreachable_code)]
    { unreachable!() }
}


pub fn not_reachable() {
    loop {}
    #[deny(unfulfilled_lint_expectations)]
    #[expect(unreachable_code)]
    { unreachable!() }
}
```


[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2c124057720ab348db8b6e02a36ae9e1)

**Comment by safinaskar** [comment]
@timotree3 , thanks a lot! I just tried to create `static_unreachable!` macro using your method, but I failed:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9bda0ef1763224a77e2f973edbabb2f6

**Comment by timotree3** [comment]
If you invoke it as `static_unreachable!()` instead of `static_unreachable!();` (remove the `;`) [it works](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7368ec0107f975d70421d2ef427149ae)

I think the `;` is making the macro expand into this
```rust
loop {}
{
    #[deny(unfulfilled_lint_expectations)]
    #[expect(unreachable_code)]
    {
        unreachable!()
    }
}; // < unreachable statement!
```

Where the outer block is an unreachable statement without any `allow` or `expect` attached to it

**Comment by timotree3** [comment]
Another way of checking that an expression diverges is to check whether you can assign it to a variable of uninhabited type. e.g. `let _: ! = loop {}`

Here is a playground for an `assert_divergent!` macro that would work on stable: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=669c7b2a6a681784d5fe95f3e06b8db8

**Comment by safinaskar** [comment]
> If you invoke it as `static_unreachable!()` instead of `static_unreachable!();` (remove the `;`) [it works](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7368ec0107f975d70421d2ef427149ae)

Thanks! This is exactly what I want. You may close this bug report

**Comment by Nilstrieb** [comment]
A `static_unreachable` macro should actually be possible to write in user code.
```rust
macro_rules! static_unreachabel {
    () => {
        #[allow(unreachable_code)]
        {
            let s = String::new();
            let _a = s;
            let _b = s;
        }
};
}
```
This should give an error when it's reachable (can't test it right now)

**Comment by safinaskar** [comment]
@Nilstrieb , thanks. Your version is better than https://github.com/rust-lang/rust/issues/107953#issuecomment-1427121026 , because your version doesn't require nightly. But it still works sub-optimally if I put `;` after macro call

**Comment by Nilstrieb** [comment]
I think you can put the insides inside an additional block to make it also work I'm expression position.

**Comment by safinaskar** [comment]
@Nilstrieb , this doesn't work

**Comment by Nilstrieb** [comment]
Ah, I misunderstood. Yeah, it doesn't mark the outer expression as reachable.

**Comment by imyxh** [comment]
+1 fixing the default behavior on this would be really nice imo. the case I just ran into where i wanted this was

```rust
loop {
    match thing() {
        Ok(thing_ok) => {
            if thing_ok.something() {
                do_something();
                i_really_need_to_do_this_before_continuing();
                continue;
            } else if thing_ok.something_else() {
                do_something_that_doesnt_terminate();
            // what if i add a bunch of messy stuff here?
            } else {
                do_another_thing();
                i_really_need_to_do_this_before_continuing();
                continue;
            }
        },
        Err(err) => {
            do_something_with_err(err);
            handle_in_a_way_that_hopefully_aborts_the_program(err);
        },
    }
    unreachable!(); // to make sure i did the thing before continuing
}
```
